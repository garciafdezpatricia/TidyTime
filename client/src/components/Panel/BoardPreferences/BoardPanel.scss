@use "../../../../styles/variables" as *;

.board-preferences {
    padding: 1rem;
    display: flex;
    flex-direction: column;
    width: calc(100% - 2rem);
    p {
        font-size: .9rem;
    }
}

.default-columns {
    font-size: .9rem;
    padding-top: .8rem;
    margin-left: 1.5rem;
    display: flex;
    flex-direction: column;
}

.columns-input {
    display: flex;
    flex-direction: column;
    width: fit-content;
    gap: .5rem;
    padding: .6rem;
    background-color: $background-button;
    border-radius: .4rem;
}

.individual-input {
    display: flex;
    gap: .5rem;

    input {
        width: 8rem;
        padding-left: .3rem;
        border-radius: .4rem;
        border: .1rem solid $light-grey;
        outline: none;
    }

    .delete-input-button {
        border: none;
        border-radius: .4rem;
        background-color: transparent;
        &:hover {
            color: $remove-red;
        }
    }
}

button.new-column {
    margin-top: .3rem;
    display: flex;
    width: fit-content;
    padding: .3rem;
    border: .1rem solid transparent;
    background-color: transparent;
    font-size: 1rem;
    &:hover {
        background-color: $primary-grey;
    }
}