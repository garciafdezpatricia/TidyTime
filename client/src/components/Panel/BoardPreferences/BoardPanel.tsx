import { useEffect, useState } from "react";
import { useTaskContext } from "../../Context/TaskContext"
import { LuListPlus } from "react-icons/lu";
import { RiDeleteBin5Fill } from "react-icons/ri";
import { useTranslation } from 'react-i18next';

export default function BoardPanel() {
    const { t } = useTranslation();
    const [columns, setColumns] = useState<string[] | undefined>(undefined);
    const { boardColumns, setBoardColumns } = useTaskContext();

    useEffect(() => {
        if (boardColumns) {
            let result:string[] = [];
            boardColumns.forEach((column) => {
                result.push(column);
            })
            setColumns(result);
        }
    }, [boardColumns])

    const handleColumnRename = (value:string, index:number) => {
        if (columns) {
            let result = [...columns];
            let updatedColumn = result[index];
            updatedColumn = value;
            result = [...result.slice(0, index),
                updatedColumn,
                ...result.slice(index + 1)
            ];
            setBoardColumns(result);
        }
    }   

    const handleColumnDeletion = (index:number) => {
        if (columns) {
            let result = columns.filter((column, columnIndex) => index !== columnIndex);
            setBoardColumns(result);
        }
    }

    const addNewInput = () => {
        if (columns) {
            let result = [...columns, "New column"];
            setBoardColumns(result);
        }
    }

    return (
        <article className="board-preferences">
            <h3>{t('preferences.board.title')}</h3>
            <hr></hr>
            <p>{t('preferences.board.desc')}</p>
            <section className="default-columns">
                {
                    columns !== undefined
                    ?
                    <div className="columns-input">
                    {
                        columns.length > 0 ?
                        columns.map((column, index) => {
                            return (
                                <div key={index} className="individual-input">
                                    <input 
                                        type="text" 
                                        defaultValue={column}
                                        onChange={(e) => handleColumnRename(e.target.value, index)}
                                    />
                                    <button  
                                        onClick={() => handleColumnDeletion(index)}
                                        className="delete-input-button">
                                        <RiDeleteBin5Fill />
                                    </button>
                                </div>
                            )
                        })
                        : <div style={{paddingInline: ".5rem"}}><p style={{fontSize: ".7rem"}}>Add a column!</p></div>
                    }
                    </div>
                    :
                    <div style={{display: "flex", alignItems: "center", gap: ".5rem"}}>
                        <p>Loading...</p>
                        <div className="loader"></div>
                    </div>
                }
                <button 
                    onClick={addNewInput}
                    className="new-column">
                    <LuListPlus />
                </button>
            </section>
        </article>
    )
}