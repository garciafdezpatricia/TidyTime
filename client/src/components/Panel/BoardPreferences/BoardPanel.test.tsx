import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import BoardPanel from '@/src/components/Panel/BoardPreferences/BoardPanel';

describe('BoardPanel', () => {
    it('renders without crashing', () => {
      const { container } = render(<BoardPanel />);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<BoardPanel />);
        expect(asFragment()).toMatchSnapshot();
      });
  });