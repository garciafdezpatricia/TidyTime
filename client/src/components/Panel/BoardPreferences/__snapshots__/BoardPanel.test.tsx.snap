// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardPanel matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="board-preferences"
  >
    <h3>
      preferences.board.title
    </h3>
    <hr />
    <p>
      preferences.board.desc
    </p>
    <section
      class="default-columns"
    >
      <div
        style="display: flex; align-items: center; gap: .5rem;"
      >
        <p>
          Loading...
        </p>
        <div
          class="loader"
        />
      </div>
      <button
        class="new-column"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 12H3"
          />
          <path
            d="M16 6H3"
          />
          <path
            d="M16 18H3"
          />
          <path
            d="M18 9v6"
          />
          <path
            d="M21 12h-6"
          />
        </svg>
      </button>
    </section>
  </article>
</DocumentFragment>
`;
